File: packages/vela-sdk/src/vela_sdk/engine/dialog_modes.py
"""Registry for dialog mode phase definitions.

Loads built-in modes from dialog_modes.yaml and allows runtime registration
of custom modes.
"""

from pathlib import Path
from typing import Optional

import yaml

from vela_sdk.schemas.workflow import DialogPhaseDefinition

_YAML_PATH = Path(__file__).parent / "dialog_modes.yaml"


class DialogModeRegistry:
    """Registry of dialog modes with lazy-loaded defaults from YAML."""

    _modes: dict[str, list[DialogPhaseDefinition]] = {}
    _loaded: bool = False

    @classmethod
    def get(cls, mode_id: str) -> Optional[list[DialogPhaseDefinition]]:
        """Return phases for a mode, or None if not registered."""
        if not cls._loaded:
            cls._load_defaults()
        return cls._modes.get(mode_id)

    @classmethod
    def register(cls, mode_id: str, phases: list[DialogPhaseDefinition]) -> None:
        """Register or override a dialog mode."""
        if not cls._loaded:
            cls._load_defaults()
        cls._modes[mode_id] = phases

    @classmethod
    def all_modes(cls) -> dict[str, list[DialogPhaseDefinition]]:
        """Return all registered modes."""
        if not cls._loaded:
            cls._load_defaults()
        return dict(cls._modes)

    @classmethod
    def _load_defaults(cls) -> None:
        """Load built-in modes from dialog_modes.yaml."""
        cls._loaded = True
        if not _YAML_PATH.exists():
            return
        with open(_YAML_PATH) as f:
            raw = yaml.safe_load(f) or {}
        for mode_id, phases_raw in raw.items():
            if not phases_raw:
                cls._modes[mode_id] = []
                continue
            cls._modes[mode_id] = [
                DialogPhaseDefinition(**p) for p in phases_raw
            ]

    @classmethod
    def _reset(cls) -> None:
        """Reset registry state (for testing)."""
        cls._modes = {}
        cls._loaded = False

File: packages/vela-sdk-ts/src/engine/dialog-modes.ts
/**
 * Registry for dialog mode phase definitions.
 *
 * Built-in modes are embedded as TypeScript objects (ported from dialog_modes.yaml).
 * Custom modes can be registered at runtime.
 */

import type { DialogPhaseDefinition } from "../schemas/workflow.js";

// ---------------------------------------------------------------------------
// Built-in modes (from dialog_modes.yaml)
// ---------------------------------------------------------------------------

const BUILTIN_MODES: Record<string, DialogPhaseDefinition[]> = {
  brainstorming: [
    {
      id: "diverge",
      name: "Divergieren",
      guideline:
        "Sammle möglichst viele Ideen ohne Bewertung. Quantität vor Qualität.",
    },
    {
      id: "converge",
      name: "Konvergieren",
      guideline:
        "Bewerte und filtere die Ideen. Welche sind realistisch und wertvoll?",
    },
    {
      id: "synthesize",
      name: "Synthese",
      guideline:
        "Fasse die besten Ideen zu einem kohärenten Ergebnis zusammen.",
    },
  ],

  requirements: [
    {
      id: "context",
      name: "Kontext",
      guideline:
        "Kläre den Hintergrund: Wer sind die Stakeholder? Was ist der Auslöser?",
    },
    {
      id: "questions",
      name: "Fragen",
      guideline:
        "Stelle gezielte Fragen um Anforderungen zu identifizieren und Lücken aufzudecken.",
    },
    {
      id: "prioritize",
      name: "Priorisieren",
      guideline:
        "Ordne die Anforderungen nach Wichtigkeit und Dringlichkeit.",
    },
    {
      id: "specify",
      name: "Spezifizieren",
      guideline:
        "Formuliere die priorisierten Anforderungen als klare, testbare Aussagen.",
    },
  ],

  planning: [
    {
      id: "goals",
      name: "Ziele",
      guideline: "Definiere klare, messbare Ziele für den Plan.",
    },
    {
      id: "breakdown",
      name: "Aufteilen",
      guideline:
        "Zerlege die Ziele in konkrete, umsetzbare Aufgaben.",
    },
    {
      id: "dependencies",
      name: "Abhängigkeiten",
      guideline:
        "Identifiziere Abhängigkeiten zwischen Aufgaben und definiere die Reihenfolge.",
    },
    {
      id: "approval",
      name: "Freigabe",
      guideline:
        "Überprüfe den Plan und bestätige oder korrigiere ihn.",
    },
  ],

  review: [
    {
      id: "understand",
      name: "Verstehen",
      guideline:
        "Stelle sicher, dass der Gegenstand der Review vollständig verstanden ist.",
    },
    {
      id: "evaluate",
      name: "Bewerten",
      guideline:
        "Bewerte systematisch nach den relevanten Kriterien.",
    },
    {
      id: "decide",
      name: "Entscheiden",
      guideline: "Triff eine klare Entscheidung mit Begründung.",
    },
  ],

  freeform: [],
};

// ---------------------------------------------------------------------------
// DialogModeRegistry
// ---------------------------------------------------------------------------

export class DialogModeRegistry {
  private static modes: Record<string, DialogPhaseDefinition[]> = {
    ...BUILTIN_MODES,
  };

  /** Return phases for a mode, or undefined if not registered. */
  static get(modeId: string): DialogPhaseDefinition[] | undefined {
    return this.modes[modeId];
  }

  /** Register or override a dialog mode. */
  static register(modeId: string, phases: DialogPhaseDefinition[]): void {
    this.modes[modeId] = phases;
  }

  /** Return all registered modes. */
  static allModes(): Record<string, DialogPhaseDefinition[]> {
    return { ...this.modes };
  }

  /** Reset to built-in modes only (for testing). */
  static _reset(): void {
    this.modes = { ...BUILTIN_MODES };
  }
}
